Update the specification with PEP 639

befeleme opened the pull request:
Changed pages:
- glossary
- pyproject.toml
- sdist spec
- wheel spec
- installation spec

:question: The meaning of the `license` key in pyproject.toml is changed. The fact that table subkeys are deprecated is mentioned. Should I also keep the previous specification details?

:question: Cosmetic "unsureness": keep the changelog at November or antidate it to August when the PEP was provisionally accepted? Since the conditions of acceptance were just recently met, it's high time for these updates.

Resolves #1245.

<!-- readthedocs-preview python-packaging-user-guide start -->
----
📚 Documentation preview 📚:
* https://python-packaging-user-guide--1662.org.readthedocs.build/en/1662/glossary/#term-License-Classifier
* https://python-packaging-user-guide--1662.org.readthedocs.build/en/1662/guides/licensing-examples-and-user-scenarios/
* https://python-packaging-user-guide--1662.org.readthedocs.build/en/1662/guides/writing-pyproject-toml/#license
* https://python-packaging-user-guide--1662.org.readthedocs.build/en/1662/specifications/binary-distribution-format/#the-dist-info-licenses-directory
* https://python-packaging-user-guide--1662.org.readthedocs.build/en/1662/specifications/pyproject-toml/#license
* https://python-packaging-user-guide--1662.org.readthedocs.build/en/1662/specifications/recording-installed-packages/#the-licenses-subdirectory
* https://python-packaging-user-guide--1662.org.readthedocs.build/en/1662/specifications/source-distribution-format/#source-distribution-file-format

<!-- readthedocs-preview python-packaging-user-guide end -->

befeleme commented:
Closes #1245

befeleme commented:
TODO: update https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#license

webknjaz commented on source/guides/licensing-examples-and-user-scenarios.rst:
```diff
@@ -0,0 +1,358 @@
+.. _chooseamitlicense: https://choosealicense.com/licenses/mit/
+.. _choosealicenselist: https://choosealicense.com/licenses/
+.. _dontchoosealicense: https://choosealicense.com/no-permission/
+.. _mitlicense: https://opensource.org/licenses/MIT
```
This generates a failure in CI: https://github.com/pypa/packaging.python.org/actions/runs/11915860405/job/33207141778?pr=1662#step:5:691. We need to either configure sphinx to stick some extra headers into requests, or add this to `linkcheck_ignore` in `conf.py`, or find a different web page to reference.

webknjaz commented on source/guides/licensing-examples-and-user-scenarios.rst:
```diff
@@ -0,0 +1,358 @@
+
+Specifically, we assume the license files are located at the following
+paths in the project source tree (relative to the project root and
+``pyproject.toml``):
```
Let's use the built-in file role for anything that can be a filesystem path:
```suggestion
:file:`pyproject.toml`):
```

webknjaz commented on source/guides/licensing-examples-and-user-scenarios.rst:
```diff
@@ -0,0 +1,358 @@
+    License-Expression: MIT
+    License-File: LICENSE
+
+The ``LICENSE`` file would be stored at ``/setuptools-${VERSION}/LICENSE``
```
Using the `:file:` role allows using just single curly braces for placeholders:
```suggestion
The :file:`LICENSE` file would be stored at :file:`/setuptools-{VERSION}/LICENSE`
```
No `$` needed. Pieces of path paths within curly braces get recognized and turned into italics. See: https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#role-file.

webknjaz commented on source/glossary.rst:
```diff
@@ -147,6 +138,49 @@ Glossary
         as described in the specicifcation :ref:`recording-installed-packages`.
 
 
+    Known Good Set (KGS)
```
Could you submit this reordering separately?

befeleme replied:
https://github.com/pypa/packaging.python.org/pull/1664

webknjaz commented on source/glossary.rst:
```diff
@@ -313,6 +347,24 @@ Glossary
+        :term:`Source Distribution (or "sdist")`;
+        and a subdirectory named ``licenses`` of
+        the directory containing the :term:`Built Metadata` —
+        i.e., the ``.dist-info/licenses`` directory —
```
```suggestion
        i.e., the :file:`.dist-info/licenses` directory —
```

webknjaz commented on source/glossary.rst:
```diff
@@ -313,6 +347,24 @@ Glossary
+        Defined to be the :term:`Project Root Directory`
+        for a :term:`Project Source Tree` or
+        :term:`Source Distribution (or "sdist")`;
+        and a subdirectory named ``licenses`` of
```
I think it's a good habit to have a trailing slash at the end of directory paths:
```suggestion
        and a subdirectory named :file:`licenses/` of
```
This is a visual clue that it's not a regular file path but a folder.

webknjaz commented on source/specifications/source-distribution-format.rst:
```diff
@@ -61,6 +61,10 @@ This directory must also contain a :file:`pyproject.toml` in the format defined
+If the metadata version is 2.4 or greater, the source distribution MUST contain
+any license files specified by the ``License-File`` field in the PKG-INFO
+at their respective paths relative to the root directory of the sdist
+(containing the :file:`pyproject.toml`` and the ``PKG-INFO`` metadata).
```
```suggestion
(containing the :file:`pyproject.toml`` and the :file:`PKG-INFO` metadata).
```

webknjaz commented:
@befeleme I've added code suggestions with `:file:` into a few places, but I'd like to ask you to also apply the same technique in other places that I've missed.

befeleme commented:
I addressed the review points in the fixup commits for better readability (will squash them later once we get closer to merging). I still walk through the pages looking for the gaps. Thank you for the sphinx tips, that's very helpful!

webknjaz commented on source/specifications/binary-distribution-format.rst:
```diff
@@ -250,6 +251,16 @@ The .dist-info directory
    mentioned and correctly hashed in RECORD.
 
 
+The .dist-info/licenses directory
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
I think the `:file:` role works in headings too
```suggestion
The :file:`.dist-info/licenses/` directory
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

webknjaz commented on source/specifications/binary-distribution-format.rst:
```diff
@@ -250,6 +251,16 @@ The .dist-info directory
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+If the metadata version is 2.4 or greater and one or more ``License-File``
+fields is specified, the .dist-info directory MUST contain a
```
```suggestion
fields is specified, the :file:`.dist-info/` directory MUST contain a
```

befeleme commented:
I *think* it's ready.

befeleme commented:
Is there anything I can do to make this move forward?

befeleme commented:
The linkcheck reports on `https://pypi.org/project/pip/23.3.1/#files` (Non-existing anchor), but that's surprising: yesterday's cron job reported is as success and the link works (opens PyPI on the files tab). Also, it's not related to any changes made for this PR.

cdce8p reviewed (commented):
Thanks for working on this @befeleme!

webknjaz commented:
> The linkcheck reports on `https://pypi.org/project/pip/23.3.1/#files` (Non-existing anchor), but that's surprising: yesterday's cron job reported is as success and the link works (opens PyPI on the files tab). Also, it's not related to any changes made for this PR.

You're right. Over the past few days I noticed a quick Fastly loading screen showing up on PyPI, which then redirects to where I was going originally.
So I probed it with cURL just now and verified that this is what's happening, and HTML DOM no longer contains that in the HTTP first response (this is probably cookie-based):
```console
$ curl -v 'https://pypi.org/project/pip/23.3.1/#files'
* Host pypi.org:443 was resolved.
* IPv6: 2a04:4e42:600::223, 2a04:4e42:200::223, 2a04:4e42:400::223, 2a04:4e42::223
* IPv4: 151.101.0.223, 151.101.192.223, 151.101.64.223, 151.101.128.223
*   Trying [2a04:4e42:600::223]:443...
* Immediate connect fail for 2a04:4e42:600::223: Network is unreachable
*   Trying [2a04:4e42:200::223]:443...
* Immediate connect fail for 2a04:4e42:200::223: Network is unreachable
*   Trying [2a04:4e42:400::223]:443...
* Immediate connect fail for 2a04:4e42:400::223: Network is unreachable
*   Trying [2a04:4e42::223]:443...
* Immediate connect fail for 2a04:4e42::223: Network is unreachable
*   Trying 151.101.0.223:443...
* ALPN: curl offers h2,http/1.1
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
*  CAfile: /etc/ssl/certs/ca-certificates.crt
*  CApath: /etc/ssl/certs
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
* TLSv1.3 (IN), TLS handshake, Finished (20):
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.3 (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / TLS_AES_128_GCM_SHA256 / x25519 / RSASSA-PSS
* ALPN: server accepted h2
* Server certificate:
*  subject: CN=pypi.org
*  start date: Apr 23 04:22:05 2024 GMT
*  expire date: May 25 04:22:04 2025 GMT
*  subjectAltName: host "pypi.org" matched cert's "pypi.org"
*  issuer: C=BE; O=GlobalSign nv-sa; CN=GlobalSign Atlas R3 DV TLS CA 2024 Q2
*  SSL certificate verify ok.
*   Certificate level 0: Public key type RSA (2048/112 Bits/secBits), signed using sha256WithRSAEncryption
*   Certificate level 1: Public key type RSA (2048/112 Bits/secBits), signed using sha256WithRSAEncryption
*   Certificate level 2: Public key type RSA (2048/112 Bits/secBits), signed using sha256WithRSAEncryption
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* Connected to pypi.org (151.101.0.223) port 443
* using HTTP/2
* [HTTP/2] [1] OPENED stream for https://pypi.org/project/pip/23.3.1/#files
* [HTTP/2] [1] [:method: GET]
* [HTTP/2] [1] [:scheme: https]
* [HTTP/2] [1] [:authority: pypi.org]
* [HTTP/2] [1] [:path: /project/pip/23.3.1/]
* [HTTP/2] [1] [user-agent: curl/8.10.1]
* [HTTP/2] [1] [accept: */*]
> GET /project/pip/23.3.1/ HTTP/2
> Host: pypi.org
> User-Agent: curl/8.10.1
> Accept: */*
> 
* Request completely sent off
< HTTP/2 200 
< set-cookie: _fs_ch_st_FSBmUei20MqUiJb9=ARwOUcLntEKxNCnL5W0on4gbZZuJgKNFAuJTwV5kqlwObPx5zOjadDLJ8iZ2jXY2v-kRpx0J1npexkvu_R75uguNU_5S13wmbTRuQr1zm4AghacYsZb2dTQG9sPxmJahlzJLe16uBKWgCnaeE4pXhqsMs77NogoTpKoqJhS6nkwgjtK2hJA3s4d8d4JnXTvMtJRqm3vtuDFWp5s6OqiT-u3N-QTbB58=; Max-Age=10; HttpOnly; Path=/
< content-type: text/html; charset=utf-8
< cache-control: no-store
< accept-ranges: bytes
< date: Wed, 11 Dec 2024 23:00:42 GMT
< x-served-by: cache-iad-kcgs7200169-IAD, cache-iad-kjyo7100141-IAD, cache-fra-eddf8230065-FRA
< x-cache: MISS, MISS
< x-cache-hits: 0, 0
< x-timer: S1733958043.869433,VS0,VE107
< strict-transport-security: max-age=31536000; includeSubDomains; preload
< x-frame-options: deny
< x-xss-protection: 1; mode=block
< x-content-type-options: nosniff
< x-permitted-cross-domain-policies: none
< permissions-policy: publickey-credentials-create=(self),publickey-credentials-get=(self),accelerometer=(),ambient-light-sensor=(),autoplay=(),battery=(),camera=(),display-capture=(),document-domain=(),encrypted-media=(),execution-while-not-rendered=(),execution-while-out-of-viewport=(),fullscreen=(),gamepad=(),geolocation=(),gyroscope=(),hid=(),identity-credentials-get=(),idle-detection=(),local-fonts=(),magnetometer=(),microphone=(),midi=(),otp-credentials=(),payment=(),picture-in-picture=(),screen-wake-lock=(),serial=(),speaker-selection=(),storage-access=(),usb=(),web-share=(),xr-spatial-tracking=()
< 
<!DOCTYPE html>
<html>
  <head>
    <meta
      http-equiv="Content-Security-Policy"
      content="default-src 'self'; img-src 'self' data:; media-src 'self' data:; object-src 'none'; style-src 'self' 'sha256-o4vzfmmUENEg4chMjjRP9EuW9ucGnGIGVdbl8d0SHQQ='; script-src 'self' 'sha256-a9bHdQGvRzDwDVzx8m+Rzw+0FHZad8L0zjtBwkxOIz4=';"
    />
    <link
      href="/_fs-ch-1T1wmsGaOgGaSxcX/assets/inter-var.woff2"
      rel="preload"
      as="font"
      type="font/woff2"
      crossorigin
    />
    <link href="/_fs-ch-1T1wmsGaOgGaSxcX/assets/styles.css" rel="stylesheet" />
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1, maximum-scale=1"
    />
    <style>
      #loading-error {
        font-size: 16px;
        font-family: 'Inter', sans-serif;
        margin-top: 10px;
        margin-left: 10px;
        display: none;
      }
    </style>
  </head>
  <body>
    <noscript>
      <div class="noscript-container">
        <div class="noscript-content">
          <img
            src="/_fs-ch-1T1wmsGaOgGaSxcX/assets/errorIcon.svg"
            alt="Error Icon"
            class="error-icon"
          />
          <span class="noscript-span"
            >JavaScript is disabled in your browser.</span
          >
          Please enable JavaScript to proceed.
        </div>
      </div>
    </noscript>
    <div id="loading-error">
      A required part of this site couldn’t load. This may be due to a browser
      extension, network issues, or browser settings. Please check your
      connection, disable any ad blockers, or try using a different browser.
    </div>
    <script>
      function loadScript(src) {
        return new Promise((resolve, reject) => {
          const script = document.createElement('script');
          script.onload = resolve;
          script.onerror = (event) => {
            console.error('Script load error event:', event);
            document.getElementById('loading-error').style.display = 'block';
            reject(
              new Error(
                `Failed to load script: ${src}, Please contact the service administrator.`
              )
            );
          };
          script.src = src;
          document.body.appendChild(script);
        });
      }

      loadScript('/_fs-ch-1T1wmsGaOgGaSxcX/errors.js')
        .then(() => {
          const script = document.createElement('script');
          script.src = '/_fs-ch-1T1wmsGaOgGaSxcX/script.js?reload=true';
          script.onerror = (event) => {
            console.error('Script load error event:', event);
            const errorMsg = new Error(
              `Failed to load script: ${script.src}. Please contact the service administrator.`
            );
            console.error(errorMsg);
            handleScriptError();
          };
          document.body.appendChild(script);
        })
        .catch((error) => {
          console.error(error);
        });
    </script>
  </body>
</html>
* Connection #0 to host pypi.org left intact
```

Nevertheless, this would be blocking PR merges, and so we have to address it by possibly adding the URL to `nitpick_ignore` or adjusting the anchor checks somehow. It's best to do this in a separate PR.

webknjaz commented on source/guides/licensing-examples-and-user-scenarios.rst:
```diff
@@ -0,0 +1,357 @@
+    classifiers =
+        License :: OSI Approved :: MIT License
+
+The simplest migration to PEP 639 would consist of using this instead:
+
+.. code-block:: ini
+
+    [metadata]
+    license_expression = MIT
```
What about those still wanting to remain on `setup.cfg`?

cc @abravalheri

befeleme replied:
> I suspect license_expression won't be added to setup.cfg.

True, this is a leftover from a previous draft implementation. `setup.cfg` now enables to define license in the `license` key. I addressed that in https://github.com/pypa/packaging.python.org/commit/f49c034b7e81f97f5015a2c5da9d17ea2c7c8107.

webknjaz commented:
> Closes #1245

@befeleme FYI such comments don't work unless they are in commit messages or the PR description.

webknjaz commented on source/guides/licensing-examples-and-user-scenarios.rst:
```diff
@@ -0,0 +1,357 @@
+paths in the project source tree (relative to the project root and
+:file:`pyproject.toml`):
+
+.. code-block:: ini
```
Why is this `ini`?
```suggestion
.. code-block:: text
```

webknjaz commented on source/guides/licensing-examples-and-user-scenarios.rst:
```diff
@@ -0,0 +1,357 @@
+the version of the Setuptools release specified in the Core Metadata,
+the license files would be located at the paths:
+
+.. code-block:: shell
```
```suggestion
.. code-block:: text
```

webknjaz commented on source/guides/licensing-examples-and-user-scenarios.rst:
```diff
@@ -0,0 +1,357 @@
+Finally, in the installed project, with :file:`site-packages/` being the site dir
+and ``{VERSION}`` as the previous, the license files would be installed to:
+
+.. code-block:: shell
```
```suggestion
.. code-block:: text
```

webknjaz commented on source/guides/licensing-examples-and-user-scenarios.rst:
```diff
@@ -0,0 +1,357 @@
+    License-Expression: LicenseRef-Custom-License
+
+
+User Scenarios
+==============
+
+User Scenarios
+--------------
```
duplicate header?

webknjaz commented on source/specifications/binary-distribution-format.rst:
```diff
@@ -182,6 +182,7 @@ its version, e.g. ``1.0.0``, consist of:
    ``purelib`` or ``platlib`` as specified in ``WHEEL``.  ``purelib`` and
    ``platlib`` are usually both ``site-packages``.
 #. ``{distribution}-{version}.dist-info/`` contains metadata.
+#. ``{distribution}-{version}.dist-info/licenses/`` contains license files.
```
Let's have explicit roles in the newly added lines (the surrounding text is out of the scope)
```suggestion
#. :file:`{distribution}-{version}.dist-info/licenses/` contains license files.
```

befeleme commented on source/guides/licensing-examples-and-user-scenarios.rst:
```diff
@@ -0,0 +1,357 @@
+organization or household, it *usually* isn't strictly necessary to include
+a formal license, so you wouldn't necessarily have to do anything extra here.
+
+However, it is still a good idea to include ``LicenseRef-Proprietary``
+as a license expression in your package configuration, and/or a
+copyright statement and any legal notices in a :file:`LICENSE.txt` file
+in the root of your project directory, which will be automatically
+included by packaging tools.
```
It may happen that package authors decide to upload a package warranted by a proprietary license to PyPI (https://pypi.org/search/?q=&o=&c=License+%3A%3A+Other%2FProprietary+License shows over 6700 packages) - this paragraph is specifically meant to cover those cases.

cdce8p replied:
> Not sure if it's the case for license expressions as well but isn't there a classifier which causes PyPI to reject uploads?

Just found it. I meant the `Private :: Do Not Upload` classifier. https://pypi.org/classifiers/

zahlman replied:
"this paragraph is specifically meant to cover those cases" - but wouldn't the package be "shared publicly" in these cases if it's uploaded to PyPI? (But I can see where it might be useful to add the designation even for proprietary code that only ever touches the company's private index - maybe it's used by the company's higher-level tooling, or is otherwise relevant to their business procedures.)

befeleme replied:
The package is shared publicly on the base of exception granted to PyPI (https://policies.python.org/pypi.org/Terms-of-Use/) but the original license applies for any consumers of that package.
Tools scanning package metadata would have a clear indicator that there's something to be aware of. An example from the field I'm familiar with: a person trying to create a Fedora rpm package out of the Python distribution would be warned that the license doesn't allow that package to be included into Fedora Linux repositories.

befeleme commented on source/guides/licensing-examples-and-user-scenarios.rst:
```diff
@@ -0,0 +1,357 @@
+included by packaging tools.
+
+
+I just want to share my own work without legal restrictions
+'''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
+
+While you aren't required to include a license, if you don't, no one has
+`any permission to download, use or improve your work <dontchoosealicense_>`__,
+so that's probably the *opposite* of what you actually want.
+The `MIT license <chooseamitlicense_>`__ is a great choice instead, as it's simple,
+widely used and allows anyone to do whatever they want with your work
+(other than sue you, which you probably also don't want).
```
The guidelines come from the original authors of the PEP whom I trust in the matter but I'm in no way a domain expert, so I don't feel competent to change recommendations. I also don't think Python packaging should be too specific when it comes to licenses. The MIT one has already appeared in https://packaging.python.org/en/latest/guides/writing-pyproject-toml/ and I'd just left it as it is.

befeleme commented on source/guides/licensing-examples-and-user-scenarios.rst:
```diff
@@ -0,0 +1,357 @@
+``license`` (``[project]`` table in :file:`pyproject.toml`),
+or the equivalent for your packaging tool,
+and make sure to remove any legacy ``license`` table subkeys or
+``License ::`` classifiers. Your existing ``license`` value may already
+be valid as one (e.g. ``MIT``, ``Apache-2.0 OR BSD-2-Clause``, etc);
```
There is the recommendation prepared by the PEP authors: https://peps.python.org/pep-0639/appendix-mapping-classifiers/#spec-mapping-classifiers-identifiers - I'd prefer to add a link to it rather than to create a new page on packaging.python.org (only a handful of classifiers can be automatically converted, it's ambiguous and "best effort") - I'd rather folks evaluated the license files of their projects again rather than just convert strings.

befeleme replied:
Addressed in: bef14b788630be828b200056f1be89b3feb194c8

zahlman commented on source/glossary.rst:
```diff
@@ -313,6 +347,24 @@ Glossary
+        Also, the root directory that their paths
+        recorded in the ``License-File``
+        :term:`Core Metadata Field` are relative to.
+        Defined to be the :term:`Project Root Directory`
```
This sentence is hard to parse because of its length and ordering. I propose separate sentences for each build artifact type, like so:
```
        For a :term:`Project Source Tree` or
        :term:`Source Distribution (or "sdist")`, this is the
        :term:`Project Root Directory`.
        For a :term:`Built Distribution` or :term:`Installed Project`,
        this is the :file:`.dist-info/licenses/` directory of
        the wheel archive or project folder respectively —
        i.e., a subdirectory named :file:`licenses/`, within the
        the directory containing the :term:`Built Metadata`.
```

(The last two lines might not actually be necessary.)

befeleme replied:
Addressed in: 839f11310e647c65609781079945dd842edc4f6e

zahlman commented on source/guides/licensing-examples-and-user-scenarios.rst:
```diff
@@ -0,0 +1,357 @@
+    ordered-set==3.1.1
+    more_itertools==8.8.0
+
+The license expressions for these projects are:
+
+.. code-block:: text
+
+    packaging: Apache-2.0 OR BSD-2-Clause
+    pyparsing: MIT
+    ordered-set: MIT
+    more_itertools: MIT
+
+A comprehensive license expression covering both Setuptools
+proper and its vendored dependencies would contain these metadata,
+combining all the license expressions into one. Such an expression might be:
+
+.. code-block:: text
+
+    MIT AND (Apache-2.0 OR BSD-2-Clause)
```
Just to be clear, figuring out this logic is supposed to be the user's responsibility, right? Not the build system's?

befeleme replied:
Yes, the specification merely provides a way to express that logic.

cdce8p replied:
Found this post in the discuss thread about it. Seems like there is a good argument that both options are valid.
https://discuss.python.org/t/pep-639-improving-license-clarity-with-better-package-metadata/2154/39

zahlman commented on source/guides/writing-pyproject-toml.rst:
```diff
@@ -324,26 +324,59 @@ You can also specify the format explicitly, like this:
+  will be matched verbatim.
+- Special characters: ``*``, ``?``, ``**`` and character ranges: [] are supported.
+- Path delimiters must be the forward slash character (``/``).
+- Patterns are relative to the directory containing :file:`pyproject.toml`, the
```
This sounds awkward. Suggestion:
```
- Patterns are relative to the directory containing :file:`pyproject.toml`, and thus may not start with a slash.
```

befeleme replied:
Addressed in: 86f049d9737815c253536cf0c33f3044891f6174

zahlman commented on source/specifications/pyproject-toml.rst:
```diff
@@ -236,16 +237,70 @@ The Python version requirements of the project.
+Tools MUST assume that license file content is valid UTF-8 encoded text,
+and SHOULD validate this and raise an error if it is not.
+
+Literal paths (e.g. :file:`LICENSE`) are treated as valid globs which means they
```
As far as I can tell, literal paths *are* valid glob patterns, not merely treated as such. Anyway, I'm not sure this use case needs to be highlighted explicitly... ?

befeleme replied:
I'll replace: `s/are treated as/are/`
The sentence comes as a result of a long discussion about the `license-files` accepted value, with the earlier PEP draft defining separate `paths` and `globs` subkeys. That draft has been implemented in some build backends. Since in the end the subkeys were "merged" into one array rather that one of them abandoned, it made sense to explicitly describe that transition. I'd keep it there for full clarity.

zahlman commented on source/specifications/pyproject-toml.rst:
```diff
@@ -236,16 +237,70 @@ The Python version requirements of the project.
+- MUST raise an error if any individual user-specified pattern does not match
+  at least one file.
+
+If the ``license-files`` key is present and
```
I'm not sure if I understood this in the PEP, either - but it seems that, when there is *no* `license-files` key, the build system can do what it wants (either omit the license files entirely, or use its own logic to choose files that correspond to the license expression, etc.)? Whatever the answer, it might be worth highlighting here.

befeleme replied:
Your assumption is correct, I'd add the proper sentence stating that.

befeleme replied:
ee95f84880d6ed240182217b2e75a6b6df9ee148

befeleme commented on source/specifications/pyproject-toml.rst:
```diff
@@ -236,16 +237,59 @@ The Python version requirements of the project.
+
+The strings MUST contain valid glob patterns, as specified below:
+
+- Alphanumeric characters, underscores (``_``), hyphens (``-``) and dots (``.``)
+  MUST be matched verbatim.
+
+- Special glob characters: ``*``, ``?``, ``**`` and character ranges: ``[]``
+  containing only the verbatim matched characters MUST be supported.
+  Within ``[...]``, the hyphen indicates a locale-agnostic range (e.g. ``a-z``,
+  order based on Unicode code points).
+  Hyphens at the start or end are matched literally.
+
+- Path delimiters MUST be the forward slash character (``/``).
+  Patterns are relative to the directory containing :file:`pyproject.toml`,
+  therefore the leading slash character MUST NOT be used.
+
+- Parent directory indicators (``..``) MUST NOT be used.
```
Addressed in: 7e12979918ec7fa9a00b99a78856fc6631ddd999

befeleme commented on source/guides/licensing-examples-and-user-scenarios.rst:
```diff
@@ -0,0 +1,357 @@
+
+    [metadata]
+    classifiers =
+        License :: OSI Approved :: MIT License
```
Addressed in: f49c034b
I moved `setup.cfg`  under the `pyproject.toml` specification for now.

webknjaz commented on source/specifications/recording-installed-packages.rst:
```diff
@@ -66,6 +66,11 @@ The ``METADATA`` file is mandatory.
 All other files may be omitted at the installing tool's discretion.
 Additional installer-specific files may be present.
 
+This ``.dist-info`` directory may contain the following directory, described in
```
```suggestion
This :file:`.dist-info/` directory may contain the following directory, described in
```

webknjaz commented on source/specifications/recording-installed-packages.rst:
```diff
@@ -66,6 +66,11 @@ The ``METADATA`` file is mandatory.
+This ``.dist-info`` directory may contain the following directory, described in
+detail below:
+
+* ``licenses/``: contains license files.
```
```suggestion
* :file:`licenses/`: contains license files.
```

webknjaz commented on source/specifications/recording-installed-packages.rst:
```diff
@@ -259,3 +275,5 @@ History
   for the full definition.
 - September 2020: Various amendments and clarifications were approved through
   :pep:`627`.
+- December 2024: The .dist-info/licenses directory was specified through
```
```suggestion
- December 2024: The :file:`.dist-info/licenses/` directory was specified through
```

webknjaz commented on source/specifications/binary-distribution-format.rst:
```diff
@@ -427,6 +438,8 @@ History
   regular files (the expected behaviour of consuming tools when encountering
   symlinks or subdirectories in this folder is not formally defined, and hence
   may vary between tools).
+- December 2024: The .dist-info/licenses directory was specified through
```
```suggestion
- December 2024: The :file:`.dist-info/licenses/` directory was specified through
```

webknjaz commented on source/guides/licensing-examples-and-user-scenarios.rst:
```diff
@@ -0,0 +1,355 @@
+either Apache 2.0 or the 2-clause BSD, your license expression would
+be ``MIT AND (Apache-2.0 OR BSD-2-Clause)``. You might have a
+:file:`LICENSE.txt` in your repo root, and a :file:`LICENSE-APACHE.txt` and
+:file:`LICENSE-BSD.txt` in the ``_vendor`` subdirectory, so to include
```
```suggestion
:file:`LICENSE-BSD.txt` in the :file:`_vendor/` subdirectory, so to include
```

webknjaz commented on source/guides/licensing-examples-and-user-scenarios.rst:
```diff
@@ -0,0 +1,355 @@
+In the built wheel, with :file:`/` being the root of the archive and
+``{VERSION}`` as the previous, the license files would be stored at:
+
+.. code-block:: shell
```
```suggestion
.. code-block:: text
```

ncoghlan reviewed (approved):
This looks good to me, so I'm taking the view that any further edits and clarifications that folks identify post merge can be tackled in a new PR.

ncoghlan commented:
Thanks for the extensive PR @befeleme, and the thorough reviews, all!

ncoghlan commented on source/guides/licensing-examples-and-user-scenarios.rst:
```diff
@@ -0,0 +1,357 @@
+all of them, you'd specify ``["LICENSE.txt", "_vendor/packaging/LICENSE*"]``
+as glob patterns, or
+``["LICENSE.txt", "_vendor/LICENSE-APACHE.txt", "_vendor/LICENSE-BSD.txt"]``
+as literal file paths.
```
I've left this topic as one to be resolved in a follow-up PR (if we decide more needs to be said)

webknjaz commented:
@ncoghlan thanks for merging! I was struggling to find time for a thorough final review myself…

ncoghlan commented:
@webknjaz I admit there was an element of "the goal is 'better than the status quo, not perfect'' in my own final review. I *almost* put it off again due to the PR size, realised other reviewers were probably doing the same thing, and decided it was time to move things forward :)

webknjaz commented:
Yeah.. I was mostly spotting low-hanging fruit in the first passes, hoping to be able to dedicate an uninterrupted chunk of time to more in-depth reading at some ideal but unspecified moment in time ;)

befeleme commented:
Thank you, all!

henryiii commented:
I think updating the guide might have been a little early, as not all backends (namely setuptools and flit-core, which are mentioned, along with others that are not mentioned like poetry-core) support PEP 639 yet. Maybe there should at least be a warning?

See https://github.com/pypa/packaging-problems/issues/870
